blockchain: Introduce DCP0007 Consensus Change

matheusd opened the pull request:
**Requires #2679**

This introduces the consensus change specificed in [DCP0007](https://github.com/decred/dcps/pull/20): Revert Maximum Treasury Expenditure Policy.

The new policy is calculated by modifying the existing function that sums treasury spends within a given block window to also return the sum of treasury adds and treasurybases. This is then used to determine the final maximum allowed expenditure according to the DCP0007 spec if the corresponding vote has passed.

rstaudt2 reviewed (commented):
This looks in line with the draft DCP from a first read-through.   Nicely done with the very organized commits, it makes the review much easier!

matheusd commented on blockchain/treasury.go:
```diff
@@ -621,24 +622,25 @@ func (b *BlockChain) sumPastTreasuryExpenditure(preTVINode *blockNode, nbBlocks
 				// amounts, so invert it here.
-				total += -v.amount
+				spent += -v.amount
+			default:
```
I think that's fair. Do you have a suggestion for handling the 'default' case? Options: log a warning, panic, return an error, do nothing.

chappjc replied:
You could also consolidate the logic with an authoritative `(*treasuryValueType).IsDebit` method, and there would be no need for a switch.  That would also reduce the number of places you need logic in the code like `wantNegative := typ == treasuryValueFee || typ == treasuryValueTSpend`, each place needing updates when new types are added.

rstaudt2 replied:
I like @chappjc's idea to have a method centralizing that logic.  I think returning an error for anything unexpected would make sense, given the function already returns an error.

matheusd replied:
Added commit https://github.com/decred/dcrd/pull/2680/commits/77011983ca6091b784e508643e72773c6f332bf7 that will be reordered to switch out existing locations to use `IsDebit()` and one to fixup the relevant commit for this PR.

matheusd commented:
Rebased against the latest #2679 and squashed the fixup commits done so far.

Note I've reordered commit https://github.com/decred/dcrd/pull/2680/commits/ea89d076e960c5b815b8a9c48052abd658af6bfd to come first in this PR.

davecgh commented on blockchain/treasury.go:
```diff
@@ -790,6 +790,13 @@ func (b *BlockChain) maxTreasuryExpenditureDCP0007(preTVINode *blockNode) (int64
 //
 // The passed node MUST correspond to a node immediately prior to a TVI block.
 func (b *BlockChain) maxTreasuryExpenditure(preTVINode *blockNode) (int64, error) {
+	isRevertPolicyActive, err := b.isRevertTreasuryPolicyActive(preTVINode)
```
I just wanted to note that both the determination of the result of the vote via this function and the expenditure calculation below are for the node _AFTER_ `preTVINode` so they line up correctly.

davecgh commented on blockchain/common_test.go:
```diff
@@ -525,6 +525,54 @@ func findDeploymentAllYesChoices(params *chaincfg.Params, voteIDs []string) (uin
+// in the voteIDs slice.
+//
+// Returns the resulting deployment version.
+func mergeAgendas(params *chaincfg.Params, voteIDs []string) (uint32, error) {
```
Nice addition!  This prevents the need to go through two voting cycles in the tests saving a lot of time.

davecgh commented on blockchain/treasury.go:
```diff
@@ -730,6 +730,58 @@ func (b *BlockChain) maxTreasuryExpenditureDCP0006(preTVINode *blockNode) (int64
+		allowedToSpend = addedPlusAllowance - spentRecent
+	}
+
+	trsyLog.Tracef("  maxTSpendExpenditureDCP0007: spent %d, "+
```
I won't hold this PR up over it, because I know it's following suit with the other logs in the treasury code, but, just noting that I really dislike the way the treasury bits in general do their logging like this.  Imo, this is something that a _caller_ should be choosing whether to log or not.  Otherwise, it ends up logging every time it's called for no reason.  Moreover, the addition of the extra spaces at the beginning make them all inconsistent with the rest of the code.

Anyway, as I said, it's not something I'll hold up this PR over, but, at some point, I'm very likely going to be ripping all these non-standard log statement out.

davecgh commented on blockchain/treasury_policy_test.go:
```diff
@@ -540,3 +551,487 @@ func TestTSpendExpendituresPolicy(t *testing.T) {
+	// Negative test: a block that attempts to spend one more than the
+	// maximum allowed by the policy is rejected.
+	//
+	// ... -> bvv# -> btspendserr2
```
Same deal here.

davecgh commented on blockchain/treasury_policy_test.go:
```diff
@@ -540,3 +551,487 @@ func TestTSpendExpendituresPolicy(t *testing.T) {
+	// Positive test: a block that attempts to spend exactly the maximum
+	// allowed by the policy is accepted.
+	//
+	// ... -> bvv# -> btspendserr2
```
Same.

davecgh reviewed (changes requested):
I'd like to point out that the overall expenditure policy test is very well done and reasonably thorough.  I know from experience that writing these complex tests when a lot of details are involved takes quite a bit of effort to get everything just right.  As far as the code change itself, it looks accurate to me and I have done some manual testing as well.

That said, I've identified a few style-related things inline as well as one issue that would cause the test to fail in the future, but otherwise this looks great.
